nixos/modules/services/tracing/tempo.nix

```diff
@@ -1,60 +1,546 @@
+      default = "local";
+      description = mdDoc ''
+        The storage backend to use.
+      '';
+    };
+
+    storageLocalPath = mkOption {
+      type = str;
+      default = "/var/lib/tempo/storage/local";
+      description = mdDoc ''
+        Where to store state if the backend selected is "local".
+      '';
+    };
+
+    storageS3Bucket = mkOption {
+      type = nullOr str;
+      default = null;
+      description = mdDoc ''
+        Bucket name in s3.  Tempo requires a dedicated bucket since it maintains a top-level
+        object structure and does not support a custom prefix to nest within a shared bucket.
+      '';
+    };
+
+    storageS3Endpoint = mkOption {
+      type = nullOr str;
+      default = null;
+      description = mdDoc ''
+        Api endpoint to connect to.  Use AWS S3 or any S3 compatible object storage endpoint.
+
+        As an example, for AWS typically this would be set to: s3.$REGION.amazonaws.com,
+        where $REGION is the region the s3 bucket was created in.
+      '';
+    };
+
+    storageS3AccessCredsEnable = mkOption {
+      type = bool;
+      default = false;
+      description = mdDoc ''
+        Whether to enable access key ENV VAR usage for static credentials.
+
+        If enabled, a file is expected to exist at location specified by option
+        storageS3AccessCredsPath.
+      '';
+    };
+
+    storageS3AccessCredsPath = mkOption {
+      type = str;
+      default = "/run/keys/tempo";
+      description = mdDoc ''
+        Specifies the location of an S3 credentials file that Tempo should utilize.
+
+        This file is expected to contain the following substituted lines:
+
+        AWS_ACCESS_KEY_ID=$SECRET_KEY_ID
+        AWS_SECRET_ACCESS_KEY=$SECRET_KEY
+      '';
+    };
+
+    storageS3ForcePathStyle = mkOption {
+      type = bool;
+      default = false;
+      description = mdDoc ''
+        Enable to use path-style requests.
+      '';
+    };
+
+    storageS3Insecure = mkOption {
+      type = bool;
+      default = false;
+      description = mdDoc ''
+        Debugging option for temporary http testing.
+      '';
+    };
+
+    storageS3InsecureSkipVerify = mkOption {
+      type = bool;
+      default = false;
+      description = mdDoc ''
+        Debugging option for temporary https testing.
+      '';
+    };
+
+    storageTraceWalPath = mkOption {
+      type = str;
+      default = "/var/lib/tempo/storage/wal";
+      description = mdDoc ''
+        Where to store the head blocks while they are being appended to.
+      '';
+    };
+
+    searchEnable = mkOption {
+      type = bool;
+      default = true;
+      description = mdDoc ''
+        Enable tempo search.
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = attrs;
+      default = {};
+      description = mdDoc ''
+        Extra configuration to pass to Tempo service.
         See https://grafana.com/docs/tempo/latest/configuration/ for available options.
       '';
```

Comment:
Like already said before me: This is a big step backwards compared to freeform settings.

Reply:
Thanks for the feedback @K900 and @SuperSandro2000.  Will review RFC42 in detail over the next few days and comment further after that.

Reply:
The PR code is from some initial work at getting a monolithic tempo service running with a few common options. A colleague expressed interest in this module to avoid having to create a working Tempo config from scratch when not that familiar with it yet and also not having much time to research the details.  I've looked over RFC42.  If I've understood correctly, it seems like I should be be able to refactor this to be more aligned with RFC42 and still provide a basic working config without much Tempo user knowledge required.  If so, then I'll do this in the near future as I can find some time.